File: python-engine/engine/deduplicator.py
from __future__ import annotations

import re
from urllib.parse import urlparse
from models.place import PlaceRecord


def _norm(s: str | None) -> str:
    return re.sub(r"[^a-z0-9]", "", (s or "").lower())


def _domain(url: str | None) -> str:
    if not url:
        return ""
    try:
        return urlparse(url).netloc.lower().removeprefix("www.")
    except Exception:
        return ""


def deduplicate(records: list[PlaceRecord]) -> list[PlaceRecord]:
    seen_ids = set()
    seen_secondary = set()
    output = []

    for r in records:
        if r.place_id and r.place_id in seen_ids:
            continue

        secondary = (
            _norm(r.business_name),
            _norm(r.phone),
            _domain(r.website),
            _norm(r.address),
        )
        # Only use secondary matching when there is meaningful overlap.
        meaningful = sum(bool(x) for x in secondary)
        if meaningful >= 2 and secondary in seen_secondary:
            continue

        if r.place_id:
            seen_ids.add(r.place_id)
        if meaningful >= 2:
            seen_secondary.add(secondary)
        output.append(r)

    return output

File: lib/data-engine/deduplicator.ts
import { PlaceRecord } from "./types";
import { normalizeWebsiteDomain } from "./normalizer";

function norm(s: string | null | undefined): string {
  return (s || "").toLowerCase().replace(/[^a-z0-9]/g, "");
}

/**
 * Global deduplication engine.
 * Priority 1: Google Place ID (place_id)
 * Priority 2: Secondary compound keys (Fuzzy Name, Phone, Domain, Address) when at least 2 fields overlap
 */
export function deduplicate(records: PlaceRecord[]): PlaceRecord[] {
  const seenIds = new Set<string>();
  const seenSecondary = new Set<string>();
  const output: PlaceRecord[] = [];

  for (const r of records) {
    if (r.place_id && seenIds.has(r.place_id)) {
      continue;
    }

    const normName = norm(r.business_name);
    const normPhone = norm(r.phone);
    const normDomain = normalizeWebsiteDomain(r.website);
    const normAddr = norm(r.address);

    const secondaryFields = [normName, normPhone, normDomain, normAddr];
    const meaningfulCount = secondaryFields.filter(Boolean).length;

    // Create a composite string key for secondary matching
    const secondaryKey = `${normName}|${normPhone}|${normDomain}|${normAddr}`;

    if (meaningfulCount >= 2 && seenSecondary.has(secondaryKey)) {
      continue;
    }

    if (r.place_id) {
      seenIds.add(r.place_id);
    }
    if (meaningfulCount >= 2) {
      seenSecondary.add(secondaryKey);
    }
    output.push(r);
  }

  return output;
}
